lib/firebase_handler.js

```diff
@@ -0,0 +1,377 @@
+          convRef.update(data);
+          resolve(true);
+        }
+      });
+    });
+  }
+
+  /**
+   * updateTimestamp - Update the lastTimestamp field of a conversation
+   * @param {string} agentId The Agent ID
+   * @param  {string} conversationId The Conversation ID
+   * @param  {int} newTime The new unix timestamp to add
+   * @return {Promise} Resolves if correctly updates timestamp
+   */
+  updateTimestamp(agentId, conversationId, newTime) {
+    return new Promise((resolve, reject) => {
+      if (agentId === undefined || conversationId === undefined ||
+          newTime === undefined) {
+        reject(new Error('Missing arguments'));
+      }
+      const db = this.database();
+      const ref = db.ref(`conversations/${agentId}/${conversationId}`);
+      // set a new timestamp in database
+      ref.once('value', (snapshot) => {
+        const data = snapshot.val();
+        data.lastTimestamp = newTime;
+        data.csatSent = false;
+        ref.update(data);
+        resolve();
+      });
+    });
+  }
+
+  /**
+   * getUserStatus - returns user status.
+   * @param {string} agentId The Agent ID
+   * @param {string} conversationId The conversation ID
+   * @param {string} displayName The displayName of conversation
+   * @return {Promise} Resolves with user status
+   */
+  getUserStatus(agentId, conversationId, displayName) {
+    return new Promise((resolve, reject) => {
+      if (agentId === undefined || conversationId === undefined ||
+          displayName === undefined) {
+        reject(new Error('Missing arguments'));
+      }
+      const db = this.database();
+      const ref = db.ref(`conversations/${agentId}/${conversationId}`);
+      // check whether user is messaging with a bot by live agent presence
+      ref.once('value', (snapshot) => {
+        const data = snapshot.val();
+        if (!data || data == {}) {
+          // create new entry with name, empty message id list, timestamp, etc.
+          ref.set({
+            name: displayName,
+            messageIds: [],
+            lastTimestamp: moment().unix(),
+            csatSent: false,
+            requestedLiveAgent: false,
+          });
+          resolve(liveAgentStatus.SPEAKING_WITH_BOT);
+        } else {
+          // check if user is currently messaging live agent or bot
+          if (data.requestedLiveAgent) {
+            resolve(liveAgentStatus.REQUESTED_LIVE_AGENT);
+          } else if (data.representativeInfo !== undefined) {
+            resolve(liveAgentStatus.SPEAKING_WITH_AGENT);
+          } else {
+            resolve(liveAgentStatus.SPEAKING_WITH_BOT);
+          }
+        }
+      });
+    });
+  }
+
+  /**
+   * saveMessage - Save a message to conversation history after it's been sent.
+   * @param  {string} agentId The agent ID
+   * @param  {string} conversationId The conversation ID
+   * @param  {Object} messageInfo The message information object
+   * @param {int} timestamp Unix time that the message was sent
```

Comment:
Is this a long?

Reply:
I think technically it should be a "number"